### PR TITLE
remove flutter update-packages from tools-tests

### PR DESCRIPTION
**Pull request by christopherfujino**
## Description

In an effort to minimize CI time spent on setup, this PR changes the shard config from calling `flutter update-packages` before each shard to instead having each shard do its own pub setup. For now, I have only updated the tool_tests to have specific `pub get` invocations, the rest of the shards still call `flutter update-packages`, but similar changes could be made to each shard, removing redundant `pub get` calls. Because `pub get` calls have been moved from the `setup` step to `main`, it is expected that the `setup` time should be significantly reduced, while time spent in `main` will be marginally increased.

This is a comparison of [this presubmit run for this PR](https://cirrus-ci.com/build/6220399350382592) to [this commit](https://cirrus-ci.com/build/4608134192824320) on master (only setup and main runtimes should be affected by this change, so I ignore all other steps):

|Version|Test|Setup Time|Main Time|Combined|
|--|--|--|--|--|
|Master|tool_tests-general-linux|2:04|1:13|3:17|
|PR|tool_tests-general-linux|0:52|1:28|2:20|
|Master|tool_tests-commands-linux|2:08|8:49|10:57|
|PR|tool_tests-commands-linux|0:55|10:03|10:58|
|Master|tool_tests-integration-linux|2:17|8:13|10:30|
|PR|tool_tests-integration-linux|1:00|11:11|12:11|
|Master|tool_tests-general-windows|4:32|1:46|6:18|
|PR|tool_tests-general-windows|2:41|2:13|4:54|
|Master|tool_tests-commands-windows|4:53|14:48|19:41|
|PR|tool_tests-commands-windows|2:10|14:51|17:01|
|Master|tool_tests-integration-windows|4:36|12:14|16:50|
|PR|tool_tests-integration-windows|2:19|12:24|14:43|
|Master|tool_tests-general-macos|2:35|1:09|3:44|
|PR|tool_tests-general-macos|1:18|1:28|2:46|
|Master|tool_tests-commands-macos|3:22|8:51|12:13|
|PR|tool_tests-commands-macos|1:19|9:53|11:12|
|Master|tool_tests-integration-macos|3:27|11:25|14:52|
|PR|tool_tests-integration-macos|1:17|11:02|12:19|

A total reduction of 6 minutes, 41 seconds.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from our [issue database]. Indicate, which of these issues are resolved or fixed by this PR. There should be at least one issue listed here.*

## Tests

I did not change any code, this is completely a CI config change to optimize test run time.

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Looks like a nice improvement to the general shard, but not as dramatic on the others (as expected, since they take longer anyway).

I would be worried about increasing the complexity of the yaml file though. Is there a way we could (using templates or some other feature) have a sort of "RUN_UPDATE_PACKGES: true" setting that we could apply by default? Then the general shards could opt out and manually run pub get

**Comment by christopherfujino** [comment]
> Looks like a nice improvement to the general shard, but not as dramatic on the others (as expected, since they take longer anyway).
> 
> I would be worried about increasing the complexity of the yaml file though. Is there a way we could (using templates or some other feature) have a sort of "RUN_UPDATE_PACKGES: true" setting that we could apply by default? Then the general shards could opt out and manually run pub get

I agree that this isn't a pure win, given the cost of complicating the config.

I'll investigate conditionally running `update-packages`

**Comment by jonahwilliams** [comment]
IT WORKS!

**Comment by jonahwilliams** [comment]
> tool_tests-general-linux Successful in 3m — Task Summary

**Comment by jmagman** [comment]
Can we pull this setup logic into the LUCI recipe?

> Looks like a nice improvement to the general shard, but not as dramatic on the others (as expected, since they take longer anyway).
> 
> I would be worried about increasing the complexity of the yaml file though. Is there a way we could (using templates or some other feature) have a sort of "RUN_UPDATE_PACKGES: true" setting that we could apply by default? Then the general shards could opt out and manually run pub get

Or instead of having a conditional variable make the `flutter update packages` line itself a template, so each task has to opt into it?  Probably ever test still remaining after the PURGE should get this treatment.

**Comment by jonahwilliams** [comment]
IIRC the framework tests occasionally call into some stuff in dev/integration_shard, it would be a bit messy.

**Comment by christopherfujino** [comment]
You basically have to remove `update-packages` from each shard, and re-run it to find all the implicit dependencies. It took me way to long just for the three tool shards.

**Comment by christopherfujino** [comment]
> You basically have to remove `update-packages` from each shard, and re-run it to find all the implicit dependencies. It took me way to long just for the three tool shards.

What we should actually do is remove `update-packages`, and make each test shard declare to `test.dart` which repos it needs initialized.

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: .cirrus.yml [hunk 1]
```diff
@@ -148,7 +149,12 @@ task:
         # CPUs, and needed at least 10G of RAM to not run out of memory.
         CPU: 4
         MEMORY: 10G
+        SHOULD_UPDATE_PACKAGES: "FALSE"
       script:
+        - (cd packages/flutter_tools; pub get)
+        - (cd packages/flutter_tools/test/data/asset_test/main; pub get)
```
Comment (jonahwilliams) [inline .cirrus.yml:155]: This is definitely something we could clean up to, by just checking in a handwritten package config.

Comment (jonahwilliams) [reply .cirrus.yml:155]: (not now)

**Comment by jmagman** [comment]
You should probably wait to rebase onto https://github.com/flutter/flutter/pull/63995.

**Comment by christopherfujino** [comment]
> You should probably wait to rebase onto #63995.

yep, I got a conflict.

**Pull request merged**